README.md
# Open-Source R Shiny Batch processing template
![GitHub release (latest by date)](https://img.shields.io/github/v/release/RRobert92/R_Shiny_Batch_processing_template)
![GitHub top language](https://img.shields.io/github/languages/top/RRobert92/R_Shiny_Batch_processing_template)
![GitHub](https://img.shields.io/github/license/RRobert92/R_Shiny_Batch_processing_template)
![GitHub Release Date](https://img.shields.io/github/release-date/RRobert92/R_Shiny_Batch_processing_template)
![GitHub contributors](https://img.shields.io/github/contributors/RRobert92/R_Shiny_Batch_processing_template)
![Status](https://img.shields.io/badge/lifecycle-experimental-orange.svg)

**Shiny Batch processing template** is an open-source template that allows for easy implementation of batch processing on multiple files with R scripts with modern GUI. 
Benefits of applying this template with your workflow:
  
1. Online access

2. Simple and intuitive to use with implemented UI. Only require installation of the R software environment and optional Rstudio.

3. It is designed from the ground up to incrementally scale up based on evolving use cases. 

4. Allows working on projects reproducible and compare data.

5. Allows to uniform and completely automate analysis.

# Table of Contents

* [Quick start](#Quick_start)
  * [Data Preparation](#Quick_start_DP)
  * [Dependency](#Dependency)
  * [Installation](#Quick_start_IN)
* [Contributing](#Contributing)
* [Copyright](#Copyright)
  
<a name="Quick_start"></a>
# Quick start
These instructions will get you a copy of the project up and running on your local machine for development and testing purposes.
You can see the real life use example of this template with in an online-version under the [Shinyapp.io](https://kiewisz.shinyapps.io/ASGA/).

<a name="Quick_start_DP"></a>
### Fork the repository and run the Shiny_template.Rproj
```
To use the template with your file format you need modifying the following files:

* Utility/Upload_Data.R <- select file format for upload and specified what to upload from the file
* Utility/Load_Data.R <- Load data for each batch round from the file format
* Utility/Check_Data.R <- Check file structure of the uploaded file to eliminate potential error during the analysis

To add any R script you need modified the following files:

* bin/tool/... <- Add your function that is needed for the analysis
* bin/Packages/Analysis/... <- Add a R script that run on loaded data with "Load_Data.R" with the function from "bin/tool/..."
* Global.R <- Source tool and packages added above
* Utility/Setting_Buttons.R <- Create new button and it reactivity for your tool. This button will be used in server.R for selecting which R script should be run 
* Server.R <- Enable newly created button under ObserveEvent(input$'Submit'){}
* Utility/Save_Data.R <- Add which data from the .Globalenviroment should be saved as a .xlsx file in the Data/... folder. This is used in Export_Data.R to collect all files and zip-them for the export.
* Utility/Report.R <- Add auto-generating plot at the end of the analysis.
```

<a name="Dependency"></a>
### Dependency for Shiny template
```
R v3.5.3 or newer
Rstudio v1.2 or newer
Java SE 11 (LTS)

R library
- shiny
- shinycssloaders
- shinyWidgets
- shinyBS
- shinyalert
- colourpicker
- readxl
- plyr
- tidyverse
- ggplot2
- egg
- base
- xlsx
- zip
```

<a name="Quick_start_IN"></a>
### Installation
```
install.packages("remotes")
remotes::install_github("RRobert92/R_Shiny_Batch_processing_templates")

```
<a name="Contributing"></a>
# Contributing
Contributions, collaborators and/or constructive criticism are welcome! Please see our Contributing Guide "Soon available" for more details.

<a href="https://sourcerer.io/rrobert92"><img src="https://avatars0.githubusercontent.com/u/56911280?v=4" height="50px" width="50px" alt=""/></a>
<a name="Copyright"></a>
## Copyright
This project is distributed under the General Public License (GPL) version 3.0 - see the [LICENSE.md](LICENSE.md) file for details.


Change_Log.txt
################################################################################
# Change_Log
#
# (c) 2019 Kiewisz
# This code is licensed under GPL V3.0 license (see LICENSE.txt for details)
#
# Author: Robert Kiewisz
# Created: 2020-10-05
################################################################################

################################################################################
# v1.00 - Unpublished
################################################################################

# Mejor changes  ---------------------------------------------------------------
- Creat Shiny template v1.00 based on the ASGA v0.32.1

# Minor changes ----------------------------------------------------------------
-
# Bug Fix ----------------------------------------------------------------------
-
# Other ------------------------------------------------------------------------
- 


tests/Test_Output.R
################################################################################
# Module Test_Result
#
# (c) 2019 Kiewisz
# This code is licensed under GPL V3.0 license (see LICENSE.txt for details)
#
# Author: Robert Kiewisz
# Created: 2020-10-05
################################################################################

# Run analysis -----------------------------------------------------------------
Run_Test <- function(input, output, session) {
  withProgress(message = "Analyzing:", value = 1, {
    incProgress(1, detail = "Test in progress...")
    Sys.sleep(0.1)

    Test_value <<- tibble()
    current_data <<- 1

    tryCatch(
      {
        callModule(Pre_Analysis, "Home") ## see/bin/Utility/Pre_Analysis.R
        Test_value[1, 1] <- TRUE
      },
      error = function(e) {
        Test_value[1, 1] <- FALSE
      }
    )
    names(Test_value)[1] <- "Pre_Analsis"

    tryCatch(
      {
        callModule(A_IKD, "Home") ## see/bin/Packages/Analysis/A_IKD.R
        Test_value[1, 2] <- TRUE
      },
      error = function(e) {
        Test_value[1, 2] <- FALSE
      }
    )
    names(Test_value)[2] <- "IKD"
    
    tryCatch(
      {
        callModule(A_MT_Bridging, "Home")
        Test_value[1, 3] <- TRUE
      },
      warning = function(w) {
        Test_value[1, 3] <- TRUE
        },
      error = function(e) {
        Test_value[1, 3] <- FALSE
      }
    )
    names(Test_value)[3] <- "MT Interaction"

    Test_value <<- Test_value
    File_name <<- as_tibble(ls(pattern = "Data_", envir = .GlobalEnv))
    File_name <<- na.omit(File_name)
  })
}

# Test analysis results --------------------------------------------------------
Test_Test <- function(input, output, session) {
  Test_df <<- data.frame()
  ## test what are the expected values for the standard data
  Test_df[1, 1] <<- nrow(File_name) == 3
  Test_df[1, 2] <<- ncol(Test_value) == length(Test_value[Test_value == TRUE])

  if (ncol(Test_df) == length((Test_df[Test_df == TRUE]))) {
    Test_df <<- TRUE
    Bad_funtions <<- 0
  } else {
    Test_df <<- FALSE
    Bad_funtions <<- which(Test_value[1, ] == FALSE, TRUE)
  }
}

# Show results  ----------------------------------------------------------------
Test_Result <- function(input, output, session) {
  ## Pop up messages after the test to confirm the results
  ## An error message is given as a number or a test which failed during the analysis
  if (Test_df == TRUE) {
    confirmSweetAlert(
      session,
      inputId = "Test_data_output",
      title = "Test result",
      text = paste("The test result was positive. All work as intended.", " Error no. ", as.character(Bad_funtions), sep = ""),
      type = "success",
      btn_labels = c("Confirm"),
      btn_colors = "#a5dc86"
    )
  } else {
    confirmSweetAlert(
      session,
      inputId = "Test_data_output",
      title = "Test result",
      text = paste("One of the test went wrong, check more infor in RStudio.
      Test results are included in `Test_value` data.frame.", "Error no. ", as.character(Bad_funtions[, 2]), sep = ""),
      type = "warning",
      btn_labels = c("Confirm"),
      btn_colors = "#f8bb86"
    )
  }
}


ui.r
################################################################################
# Shiny UI-Home
#
# (c) 2019 Kiewisz
# This code is licensed under GPL V3.0 license (see LICENSE.txt for details)
#
# Author: Robert Kiewisz
# Created: 2020-10-05
################################################################################

# Shiny UI-Home  ---------------------------------------------------------------
fluidPage(
  includeCSS("www/css/style.css"),

  useShinyalert(),

  navbarPage(
    title = App_title, # see Global.R
    collapsible = TRUE,
    inverse = TRUE,
    position = "fixed-top",
    id = "innavbar",
    selected = "Home",
    # footer = footnoteUI("footnote"),
    tabPanel(
      "GetStarted",
      GetStarted_UI("Home")
    ),

    tabPanel(
      "Home",
      fluidRow(
        homeUI("Home")
      )
    ),

    tabPanel(
      "Wiki",
      fluidRow()
    ),

    tabPanel(
      "About",
      fluidRow()
    )
  )
)


global.r
################################################################################
# Shiny Global
#
# (c) 2019 Kiewisz
# This code is licensed under GPL V3.0 license (see LICENSE.txt for details)
#
# Author: Robert Kiewisz
# Created: 2020-10-05
################################################################################
# Title of the app -------------------------------------------------------------
App_title <- "Shiny batch template"

# Global CSS  ------------------------------------------------------------------
includeCSS("www/css/style.css")

# Global HTML  -----------------------------------------------------------------
source("www/Home/index.R")
source("www/Get_Started/index.R")

# Global Utility  --------------------------------------------------------------
source("bin/Utility/Library.R") ## load all necessary library 
source("bin/Utility/Check_Data.R") ## check if data are in the right format
source("bin/Utility/Upload_Data.R") ## upload data in .xlsx or any other format
source("bin/Utility/Setting_Buttons.R") ## set-up the button for selecing analysis
source("bin/Utility/Pre_Analysis.R") ## standarise data befor analysis (sorting)
source("bin/Utility/Load_Data.R") ## Load uploaded data into the enviroment
source("bin/Utility/Save_Data.R") ## save .xlsx data from Data/... 

# Global server settings  ------------------------------------------------------
options(shiny.maxRequestSize = 1024 * 1024^2)
options(shiny.host = "127.0.0.1")
options(shiny.port = 7878)

# Global tool settings  --------------------------------------------------------
DataTest <<- 0
numfiles <<- 0
Minus_Distance <<- 0.035 # Minus end distance of any MT to the KMT [um]
MT_point_config <<- 0.035 # Distance of any MT to the MT [um]

# Global Functions for pre analysis --------------------------------------------
source("bin/Tools/Pre_Analysis/Sort_by_Fiber.R")
source("bin/Tools/Pre_Analysis/Select_Points.R")
source("bin/Tools/Pre_Analysis/Find_XYZ.R")
source("bin/Tools/Pre_Analysis/Kinetochore_Position.R")
source("bin/Tools/Pre_Analysis/Sort_All_Points_to_Start_From_the_Kinetochore.R")
source("bin/Tools/Pre_Analysis/T_Relative_Position.R")
source("bin/Tools/Pre_Analysis/Length_Distiribution.R")

# Global Functions for analysis ------------------------------------------------
source("bin/Tools/Analysis/T_Inter_Kinetochore_Dist.R")
source("bin/Tools/Analysis/T_Fiber_Area.R")
source("bin/Tools/Analysis/T_Bridging_MT.R")

source("bin/Utility/Report.R")

# Global Packages  -------------------------------------------------------------
source("bin/Packages/Analysis/A_IKD.R")
source("bin/Packages/Analysis/A_Fiber_Area.R")
source("bin/Packages/Analysis/A_MT_Bridging.R")

# Test Unit ---------------------------------------------------------------------
source("tests/Test_Output.R")

www/Home/index.R
################################################################################
# Shiny UI-Home
#
# (c) 2019 Kiewisz
# This code is licensed under GPL V3.0 license (see LICENSE.txt for details)
#
# Author: Robert Kiewisz
# Created: 2020-10-05
################################################################################

# Shiny UI-Home  ---------------------------------------------------------------
  ## Home page seen after the shiny is started ---------------------------------

homeUI <- function(id) {
  ns <- NS(id)

  tags$div(class = "splash-container",
    tags$div(class = "splash",
      tags$h1(class = "splash-head",
        "Shiny Batch processing template"
      ),
      tags$p(class = "splash-subhead",
        "The template intended to allow for easy running the R script on multiple files wrap up in modern UI."
      ),
      
      actionButton("GetStarted", "Get Started", class = "asga-button asga-button-primary"),
      actionButton("Wiki", "Wiki", class = "asga-button asga-button-primary")
    ),
    
    tags$div(class = "footer l-box is-center",
      tags$p("© Copyright GPL V3.0 2019, Robert Kiewisz")
    )
  )
}

# UI-Footnote  ---------------------------------------------------------------

footnoteUI <- function(id) {
  ns <- NS(id)

  tags$div(class = "footer l-box is-center",
    tags$p("© Copyright GPL V3.0 2019, Robert Kiewisz")
  )
}


server.r
################################################################################
# Shiny Server
#
# (c) 2019 Kiewisz
# This code is licensed under GPL V3.0 license (see LICENSE.txt for details)
#
# Author: Robert Kiewisz
# Created: 2020-10-05
################################################################################

# Shiny Server  ----------------------------------------------------------------
function(input, output, session) {

  # Hide pages  ----------------------------------------------------------------
  hideTab(inputId = "innavbar", target = "GetStarted")
  hideTab(inputId = "innavbar-GS", target = "Settings")
  hideTab(inputId = "innavbar-GS", target = "Report")

  # Get_Started button  --------------------------------------------------------
  observeEvent(input$GetStarted, {
    if (numfiles == 0) {
      updateTabsetPanel(session, "innavbar", selected = "GetStarted")
      showTab(inputId = "innavbar", target = "GetStarted")
      updateTabsetPanel(session, "innavbar-GS", selected = "UploadData")
    } else {
      updateTabsetPanel(session, "innavbar", selected = "GetStarted")
      showTab(inputId = "innavbar", target = "GetStarted")
      updateTabsetPanel(session, "innavbar-GS", selected = "Settings")
    }
  })

  # Wiki button  ---------------------------------------------------------------
  observeEvent(input$Wiki, {
    updateTabsetPanel(session, "innavbar", selected = "Wiki")
    hideTab(inputId = "innavbar", target = "GetStarted")
  })

  # Get file and Load data  ----------------------------------------------------
  callModule(Getfiles_Server, "Home")

  # Upload data UI  ------------------------------------------------------------
  output$Upload <- renderUI({
    UploadData_UI("GetStarted")
  })

  # Load standard data ---------------------------------------------------------
  observeEvent(input$`Test_unit`, {
    showTab(inputId = "innavbar-GS", target = "Settings")
    updateTabsetPanel(session, "innavbar-GS", selected = "Settings")

    numfiles <<- 1
    DataTest <<- 1
    Test <<- TRUE
  })

  # Download zip files with analyzed data --------------------------------------
  output$downloadData <- downloadHandler(
    filename = function() {
      "ASGA_Data.zip"
    },
    content = function(fname) {
      setwd("Data/")
      on.exit(setwd("../"))

      Zip_Files <- list.files(path = getwd(), pattern = ".xlsx$")
      zipr(zipfile = fname, files = Zip_Files)

      file.remove(Zip_Files)
    }
  )

  # Page responsiveness after loading data  ------------------------------------
  observeEvent(input$`Home-file`, {
    Test <<- FALSE

    showTab(inputId = "innavbar-GS", target = "Settings")
    if (DataTest == 1) {
      updateTabsetPanel(session, "innavbar-GS", selected = "Settings")
    } else {
      updateTabsetPanel(session, "innavbar", selected = "Home")
    }
  })

  # Page responsiveness after analyzing data  ----------------------------------
    ## This part collect information about all analyzed data and based on that 
    ## will generate plot. This part is only used when analyzed data were uploaded
  observeEvent(input$`Home-file1`, {
    Test <<- FALSE

    showTab(inputId = "innavbar-GS", target = "Report")
    updateTabsetPanel(session, "innavbar-GS", selected = "Report")

    File_name <<- as.data.frame(File_name)
    numfiles <<- readr::parse_number(File_name[nrow(File_name), 1])

    df <- data.frame()

    for (i in 1:nrow(File_name)) {
      name <- as.data.frame(str_split(File_name[i, 1], "_"))
      df[i, 1] <- as.numeric(name[2, 1])
      name <- as.data.frame(str_split(File_name[i, 1], paste("Data_", df[i, 1], "_", sep = "")))
      df[i, 2] <- as.character(name[2, 1])
    }

    File_name <<- df

    rm(df, name)

    # Collect information to start a plot after analysis -----------------------

    lapply(1:numfiles, function(i) {
      observeEvent(input[[paste("Data_label", i, sep = "_")]], {
        assign(paste("Data_label", i, sep = "_"),
          input[[paste("Data_label", i, sep = "_")]],
          envir = .GlobalEnv
        )
      })

      observeEvent(input[[paste("Data_color", i, sep = "_")]], {
        assign(paste("Data_color", i, sep = "_"),
          input[[paste("Data_color", i, sep = "_")]],
          envir = .GlobalEnv
        )
      })
    })

    callModule(Report_Plot_Settings, "Home") ## see bin/Utility/Report.R
    callModule(Report_Plot, "Home") ## see bin/Utility/Report.R
  })

  # Reactivity for the Home and GS button  -------------------------------------
  observe({
    if (req(input$`innavbar-GS`) == "Home") {
      updateTabsetPanel(session, "innavbar", selected = "Home")
      hideTab(inputId = "innavbar", target = "GetStarted")
    } else if (req(input$`innavbar-GS`) == "UploadData") {
      updateTabsetPanel(session, "innavbar-GS", selected = "UploadData")
    } else if (req(input$`innavbar-GS`) == "Settings") {
      updateTabsetPanel(session, "innavbar-GS", selected = "Settings")
    }
  })

  # Reactivity for the Settings button  ----------------------------------------
  callModule(Setting_Buttons_Server, "Home") ## see bin/Utility/Setting_Buttons.R

  # Reactivity for Pre-Analysis  -----------------------------------------------
   ## This part running the R code based on user selection for each uploaded file
  
  observeEvent(input$`Submit`, {
    if (Test == FALSE) {
      withProgress(message = "Analyzing:", value = 1, {
        for (y in 1:numfiles) {
          current_data <<- y
          incProgress(1 / numfiles, detail = paste("Data set no.", y, sep = " "))
          Sys.sleep(0.1)

          callModule(Load_Data, "Home")      ## see bin/Utility/Pre_Analysis.R
          callModule(Pre_Analysis, "Home")

          if (input$`Home-All_Anaysis` == TRUE) {
            callModule(A_IKD, "Home") ## see bin/Packages/Analysis/A_IKD.R
            callModule(A_Fiber_Area, "Home") ## see bin/Packages/Analysis/A_Fiber_Area.R
            callModule(A_MT_Bridging, "Home")## see bin/Packages/Analysis/_MT_Bridging.R
          }

          if (input$`Home-KMT_number` == TRUE) {
            callModule(A_KMT_number, "Home") ## see bin/Packages/Analysis/A_KMT_number.R
          }
          
          if (input$`Home-KMT_number` == TRUE) {
            callModule(A_Fiber_Area, "Home") ## see bin/Packages/Analysis/A_Fiber_Area.R
          }
          
          if (input$`Home-MT_Interaction` == TRUE) {
            callModule(A_MT_Bridging, "Home")## see bin/Packages/Analysis/_MT_Bridging.R
          }

          callModule(Save_Data, "Home")      ## see bin/Utility/Save_Data.R
        }

        showTab(inputId = "innavbar-GS", target = "Report")
        updateTabsetPanel(session, "innavbar", selected = "Report")

        ## This part collect information about all analyzed data and based on that 
        ## will generate plot.
        File_name <<- as.data.frame(ls(pattern = "Data_", envir = .GlobalEnv))
        numfiles <<- readr::parse_number(File_name[nrow(File_name), 1])

        df <- data.frame()

        for (i in 1:nrow(File_name)) {
          name <- as.data.frame(str_split(File_name[i, 1], "_"))
          df[i, 1] <- as.numeric(name[2, 1])
          name <- as.data.frame(str_split(File_name[i, 1], paste("Data_", df[i, 1], "_", sep = "")))
          df[i, 2] <- as.character(name[2, 1])
        }

        File_name <<- na.omit(df)

        rm(df, name)
      })

      # Download data ----------------------------------------------------------
      output$`Home-Download_Button` <- renderUI({
        downloadBttn("downloadData", label = "Download", style = "material-flat", color = "success")
      })

      # Collect information to start a plot after analysis ---------------------
      lapply(1:numfiles, function(i) {
        observeEvent(input[[paste("Data_label", i, sep = "_")]], {
          assign(paste("Data_label", i, sep = "_"),
            input[[paste("Data_label", i, sep = "_")]],
            envir = .GlobalEnv
          )
        })

        observeEvent(input[[paste("Data_color", i, sep = "_")]], {
          assign(paste("Data_color", i, sep = "_"),
            input[[paste("Data_color", i, sep = "_")]],
            envir = .GlobalEnv
          )
        })
      })

      callModule(Report_Plot, "Home") ## see bin/Utility/Report.R
    } else {
      ## Load data for test unit. In the example case excel file has 3 sheets
      Nodes <<- read_excel(
        "tests/ASGA_Test_Data_Set.xlsx",
        sheet = "Nodes"
      )

      Segments <<- read_excel(
        "tests/ASGA_Test_Data_Set.xlsx",
        sheet = "Segments"
      )

      Points <<- read_excel(
        "tests/ASGA_Test_Data_Set.xlsx",
        sheet = "Points"
      )
      
      ## Run test see tests/Test_Output.R
      callModule(Load_Data, "Home")   ## see tests/Test_Output.R
      callModule(Run_Test, "Home")    ## see tests/Test_Output.R
      callModule(Test_Test, "Home")   ## see tests/Test_Output.R

      updateTabsetPanel(session, "innavbar-GS", selected = "UploadData")
      callModule(Test_Result, "Home") ## see tests/Test_Output.R
      
      ## Clean data/...
      setwd("Data/")
      Files <<- list.files(path = getwd(), pattern = ".xlsx$")
      file.remove(Files)
      setwd("../")
      
      ## if test pass restore environment to the beginning
      if (Test_df == TRUE) {
        rm(list = ls(envir = .GlobalEnv), envir = .GlobalEnv)
      }

      Test <<- FALSE
      source("global.r")
    }
  })

  # Report page output ---------------------------------------------------------
  output$`Home-Plot_Settings` <- renderUI({
    Report_Plot_Settings("Report") ## see bin/Utility/Report.R
  })

  output$`Home-Report_Page` <- renderUI({
    tagList(
      if (length(File_name[File_name$V2 == "KMT_No", 2]) >= 1) {
        tagList(
          tags$p(class = "splash-subhead-Report", "KMTs number per kinetochore"),

          Report_Plot_KMT_No("Report")
        )
      }
    )
  })

  # Refresh for the Report page ------------------------------------------------
  observeEvent(input$Refresh, {
    callModule(Report_Plot, "Home") ## see bin/Utility/Report.R
  })
}
